rust-runtime/aws-smithy-http-server-python/src/util.rs

```diff
@@ -42,8 +42,45 @@ fn is_coroutine(py: Python, func: &PyObject) -> PyResult<bool> {
         .extract::<bool>()
 }
 
+// Checks whether given Python type is `Optional[T]`.
+pub fn is_optional_of<T: PyTypeInfo>(py: Python, ty: &PyAny) -> PyResult<bool> {
+    // for reference: https://stackoverflow.com/a/56833826
+
+    // in Python `Optional[T]` is an alias for `Union[T, None]`
+    // so we should check if the type origin is `Union`
+    let union_ty = py.import("typing")?.getattr("Union")?;
+    match ty.getattr("__origin__").map(|origin| origin.is(union_ty)) {
+        Ok(true) => {}
+        // Here we can ignore errors because `__origin__` is not present on all types
+        // and it is not really an error, it is just a type we don't expect
+        _ => return Ok(false),
+    };
+
+    let none = py.None();
+    // in typing, `None` is a special case and it is converted to `type(None)`,
+    // so we are getting type of `None` here to match
+    let none_ty = none.as_ref(py).get_type();
+    let target_ty = py.get_type::<T>();
+
+    // `Union` should be tuple of `(T, NoneType)`
```

Comment:
Should the edge case of `None | T` be considered? The convention right now is `T | None` absolutely, but it could produce some confusing errors otherwise.